### DISCUSSION TITLE
Labels logic per MR

### DISCUSSION THREAD

**Discussion by cdsre** [renovatebot/renovate#27357] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

16.62.10

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### What is your question?

I have set up presets to allow us to use modular config for composition. In this use case I have a devops config and a terraform config. My terraform config is meant to split the update cadence for internal modules and external. This works as expected. However I wanted to apply labels in a dynamic way such that the devops label is set by the devops config. and then the terraform config will add additional labels if rules match. 

**devops.json**
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended",
    "local>devops/renovate-runner//preset_configs/terraform#renovate-test-DEVOPSS-1091",

  ],
  "printConfig": true,
  "configMigration": true,
  "groupName": "all dependencies",
  "groupSlug": "all",
  "lockFileMaintenance": {
    "enabled": false
  },
  "labels": ["DevOps"],
  "packageRules": [
    {
      "groupName": "all dependencies (Internal)",
      "matchPackagePrefixes": [
        "mycompany/",
        "com.mycompany"
      ]
    }
  ]
}
```

**terraform.json**
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "packageRules": [
    {
      "groupName": "all dependencies tf (External)",
      "description": "This catches all other terraform updates",
      "matchManagers": [
        "terraform",
        "terraform-version"
      ],
      "addLabels": [
        "terraform",
        "internal"
      ],
      "excludePackagePrefixes": [
        "https://gitlab.mycompany.com/"
      ],
      "extends": [
        "schedule:weekly"
      ]
    },
    {
      "groupName": "all dependencies tf (Internal)",
      "description": "This should catch internal terraform modules that are in our own gitlab.",
      "matchManagers": [
        "terraform"
      ],
      "addLabels": [
        "terraform",
        "external"
      ],
      "matchPackagePrefixes": [
        "https://gitlab.mycompany.com/"
      ],
      "schedule": [
        "at any time"
      ]
    }
  ]
}
```

When I run renovate it correctly groups these and doesnt schedule the external ones every run.

```
{
           "branchName": "renovate/all-dependencies-tf-(external)",
           "prNo": null,
           "prTitle": "Update all dependencies tf (External)",
           "result": "not-scheduled",
           "upgrades": [
             {
               "datasource": "terraform-provider",
               "depName": "aws",
               "displayPending": "",
               "fixedVersion": "5.0.0",
               "currentVersion": "5.0.0",
               "currentValue": "=5.0.0",
               "newValue": "5.37.0",
               "newVersion": "5.37.0",
               "packageFile": "versions.tf",
               "updateType": "minor",
               "packageName": "hashicorp/aws"
             },
             {
               "datasource": "terraform-module",
               "depName": "terraform-aws-modules/vpc/aws",
               "displayPending": "",
               "fixedVersion": "5.0.0",
               "currentVersion": "5.0.0",
               "currentValue": "5.0.0",
               "newValue": "5.5.2",
               "newVersion": "5.5.2",
               "packageFile": "versions.tf",
               "updateType": "minor",
               "packageName": "terraform-aws-modules/vpc/aws"
             }
           ]
         },
         {
           "branchName": "renovate/all-dependencies-tf-(internal)",
           "prNo": null,
           "prTitle": "Update all dependencies tf (Internal)",
           "result": "done",
           "upgrades": [
             {
               "datasource": "git-tags",
               "depName": "gitlab.mycompany.com/devops-terraform-modules/iot-core",
               "displayPending": "",
               "fixedVersion": "1.5.0",
               "currentVersion": "1.5.0",
               "currentValue": "1.5.0",
               "newValue": "1.7.1",
               "newVersion": "1.7.1",
               "packageFile": "versions.tf",
               "updateType": "minor",
               "packageName": "https://gitlab.mycompany.com/devops-terraform-modules/iot-core"
             },
             {
               "datasource": "git-tags",
               "depName": "gitlab.mycompany.com/devops/consul-ami-builds",
               "displayPending": "",
               "fixedVersion": "4.0.0",
               "currentVersion": "4.0.0",
               "currentValue": "4.0.0",
               "newValue": "4.7.0",
               "newVersion": "4.7.0",
               "packageFile": "versions.tf",
               "updateType": "minor",
               "packageName": "https://gitlab.mycompany.com/devops/consul-ami-builds"
             }
           ]
         }
```
However when the internal MR comes it has the tags from the external tf as well 
![image](https://github.com/renovatebot/renovate/assets/16945936/4a06e9c5-016d-43c5-9dc5-ef14d01a01e0)

I suspect this is because the external tf rule was matched but just wasn't scheduled. Is there a way to set labels at the grouping level so i can have different labels for my tf-internal slug and my tf-external slug. My initial plan was to add more presets like this and I would want my internal things to be tagged only with labels that were present in the devops config or in the rule that applied to this group.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment]
This doesn't seem right: 
<img width="280" alt="image" src="https://github.com/renovatebot/renovate/assets/6311784/4bdbad66-7d38-4fa3-89ce-6fdf4b1458c7">

**Reply by cdsre** [reply]
Spot on, I seem to have the tags the wrong way around, but regardless of that I think the question still stands which is essentially I want the MR to have only the tags of the rules which matched that are in the MR and not the rules which are matched but not scheduled.

I.E I don't want to get `internal` and `external` on the same MR since they will always come in separate MR's. I am not sure if renovate allows logic to apply tags to specific groupings rather than just to any MR created in that renovate run on the project.

**Reply by rarkins** [reply]
If you have the tags from both then it means that somehow both matched

**Reply by cdsre** [reply]
Yeah I suspect its because both do match technically since I have both internal and external version upgrade. The internal ones get put into the tf-internal slug and the external ones get put into the tf-external slug. The internal one is then created and the external one is not schedules which is what I expect. 

I guess what I am trying to determine is if tags are generated for all rule matches and then applied to any MR's regardless of its group slug?

**Reply by cdsre** [reply]
Essentially what I was hoping to achieve is to have different sets of tags for each group slug. Since I want to try to build these modularly and not  have labels from one preset be present on labels from another preset which creates its own MR.

I know I can redefine labels at the group level but I would then lose any that were defined in higher preset like my devops.json. 

So ideally I would like to end up with for example two MRs with the labels like

tf-internal - devops, terraform, internal
tf-external - devops, terraform, external

but if some other team had their own config but used the terraform preset, they might end up with 

tf-internal - sre, terraform, internal
tf-external - sre, terraform, external

So i want to avoid hardcoding or hard overwriting the labels in the package rule instead i was hoping addlabels would apply at the group level not at the project level to any MR.
